File: replace.go
package main

import "fmt"
import "strings"
import "regexp"

func interpreter(program string, debug bool, separator string) string {
	var code = program;
	var lines = strings.Split(code, "\n")
	var regex = regexp.MustCompile("(.*)" + separator + "(.*)")

	if(debug) {
		fmt.Println("0:")
		fmt.Println(code)
	}

	for i := 0; i < len(lines); i++ {
		var line = lines[i]

		if regex.MatchString(line) {
			var groups = regex.FindAllStringSubmatch(line, -1)[0]
			var find = groups[1];
			var replace = groups[2];

			var rfind = regexp.MustCompile(find)
			code = rfind.ReplaceAllString(code, replace)

			if(debug){
				fmt.Print(i+1)
				fmt.Println(":", find, separator, replace)
				fmt.Println(code)
			}

			lines = strings.Split(code, "\n")
		}
	}
	return code;
}

func main() {
	fmt.Println(interpreter("(.*\\n)/$1$1\n", true, "/")) //infinit loop
}

File: replace.kt
fun interpreter(code: String, debug: Boolean, separator: String) {
	var code = code
	var lines = code.split("\n")
	var regex = ("(.*)" + separator + "(.*)").toRegex()
	var i = 0

	if(debug) {
		println(i.toString() + ": ")
		println(code)
	}

	while(i < lines.size) {
		var line = lines.get(i++)
		if(regex.matches(line)) {
			var groups = regex.matchEntire(line)!!.groups
			var find = groups.get(1)!!.value
			var replace = groups.get(2)!!.value
			var rfind = find.toRegex()

			code = rfind.replace(code, replace)
			if(debug) {
				println(i.toString() + ": " + find + " " + separator + " " + replace)
				println(code)
			}
			lines = code.split("\n")
		}
	}
}

fun main(args: Array<String>) {
	interpreter("(.*)/$1$1", true, "/")
}